Fix python

NathanielHill opened the pull request:
I think this is all that's needed. May need to make sure the encoding is correct when passing the response back out.

@paulogdm
@matheuss

NathanielHill commented:
I'm just spitballing on the `requirements.txt` fix, but seemed like it's helpful to get a PR started.

I'm wondering if it's a good idea to be installing the `requests` package without an explicit version? Seems like a recipe for a failure down the line if they ever release breaking changes.

TooTallNate commented on packages/now-python/now_handler.py:
```diff
@@ -12,8 +12,10 @@ def now_handler(event, context):
     path = payload['path']
     headers = payload['headers']
     method = payload['method']
+    data = payload['body']
```
Note that there is also a `encoding` field when `body` is set, and it _may_ be set to "base64". The handler should probably take care of decoding the data before invoking the user code.

timneutkens replied:
Does this change relate to the requirements.txt change 🤔

NathanielHill replied:
@timneutkens No, they are not related other than the fact that they are major blocking problems preventing me from deploying a Python endpoint. I can do separate PRs

NathanielHill replied:
@TooTallNate What are the possible values of `encoding`? And that's `payload['encoding']`? The nested `body` is confusing.
Is there any easy way to test/debug builders without having to publish to npm?

NathanielHill replied:
As it stands, an empty request body is also a problem:

```
Traceback (most recent call last):
  File "/var/task/now__handler__python.py", line 15, in now_handler
    data = payload['body']
KeyError: 'body'
```

darknoon commented:
@TooTallNate is there any way review of this could get pushed forward? I tried building a project on Now V2 that needs to run some python code against POST data. Would be simple to use Now V1.

Alternately, is there a way I could test out a build of this branch? It's not clear to me how to test custom builders on Now V2. Do I need to push my code to NPM?

TooTallNate commented:
The diff seems to be only related to passing the request body in to the Python server. What happened to the `requirements.txt` changes?

NathanielHill commented:
@darknoon Yes, I published my own builder to NPM to test it out. I'm actually using that builder in production now until this gets resolved by zeit. Would be nice if there was a simpler workflow to write/test builders.

@TooTallNate The `requirements.txt` change I made wasn't working, so I removed it and figured it should be it's own PR anyways

NathanielHill commented:
@TooTallNate I've only got one Python Lambda at the moment, so no `requirements.txt` conflict for me - although it's ugly. I also don't have any encoding trouble getting JSON in and out of my lambda with this diff.

I don't have time to devote to those two fixes right now, so someone else will have to take care of that to merge this.

NathanielHill commented:
Okay, I've added `base64` encoding support, and added local `requirements.txt` support as well.

I've left in support for a global `requirements.txt` in monorepos. This seems wrong, but I don't want to introduce a breaking change as people may be relying on the behavior now. Feel free to change that :+1: 

There are currently 8 Issues/PRs about improving Python build support. This should take care of the basic stuff and perhaps the remaining WSGI/framework PRs can be merged into one so progress can be made there? I'm just using a custom python `BaseHTTPRequestHandler` wrapper, so I have no need for that, hopefully others can pick up the ball on that.

#14 #26 #27 #87 #92 #95 #163 

This PR will directly fix #87 and #92 once merged.

@TooTallNate @timneutkens @matheuss @rauchg

NathanielHill commented:
Had a quick look at the CI logs and there appears to be a lot of breakage unrelated to this PR :man_shrugging:. Let me know if there's anything I need to take care of.

Contextualist commented on packages/now-python/now_handler.py:
```diff
@@ -12,12 +12,20 @@ def now_handler(event, context):
+        body = base64.b64decode(body)
+
+    res = requests.request(method, 'http://0.0.0.0:3000' + path,
+            headers=headers, data=body)
```
```suggestion
            headers=headers, data=body, allow_redirects=False)
```

It's probably a better idea to disable server-side redirection and leave the response (e.g. 302 Found) as it is, or else the redirection will trigger an error in Now's execution environment.

See kenneth-reitz/now-builders#1 for details.

NathanielHill replied:
I'm good with this change if the team approves

TooTallNate replied:
Makes sense to me 👍

NathanielHill commented:
@tootallnate ready to go?

NathanielHill commented:
In the long run, a lambda using a `requirements.txt` somewhere else in a monorepo is unexpected behavior that should probably be removed.

TooTallNate commented on packages/now-python/now_handler.py:
```diff
@@ -12,12 +12,20 @@ def now_handler(event, context):
     method = payload['method']
-    
-    res = requests.request(method, 'http://0.0.0.0:3000' + path, headers=headers)
+    encoding = payload.get("encoding")
```
I'm not all that familiar with Python, but why using `.get()` here instead of the bracket syntax?

NathanielHill replied:
Also not a regular Python user. I got the `.get()` method usage from another snippet trying to solve this problem. Difference is, itt does not appear to trigger a key not found error if the key does not exist on the object. Since `encoding` is optional, we need to avoid an error if it doesn't exist.

NathanielHill commented on packages/now-python/index.js:
```diff
@@ -40,8 +40,16 @@ exports.build = async ({ files, entrypoint }) => {
+    await pipInstall(pipPath, srcDir, '-r', requirementsTxtPath);
+  } else if (files['requirements.txt']) {
+    console.log('found global "requirements.txt"');
 
```
@timneutkens Does this builder logic at least look good?

NathanielHill commented:
Maybe @matheuss knows Python :grimacing: :man_shrugging:

paulogdm commented:
Hey people!
Not a heavy python user, but we should consider to add custom `pip` flags for a number of reasons: https://github.com/pypa/pip/issues/5739

NathanielHill commented:
@paulogdm I like that idea. Maybe we can add that to it's own PR or tack it on to one of the WSGI PRs.

Hoping this can get merged soon and the builder can be functional for simple projects.

honzajavorek commented:
Hi, I was subscribed under this issue and I noticed multiple "I don't have python experience" comments. I don't know Now builders that well, but I'm fluent in both JavaScript and Python. I'm happy to help with reviews, if you want.

NathanielHill commented:
@honzajavorek It's already merged, but feel free to have a look at `now_handler.py` and make sure everything looks legit :+1:

honzajavorek commented:
Yup, originally I wanted to approve the PR, but it got merged under my hands while I was writing the comment 😄 

I noticed only a few coding style nitpicks, which I wouldn't point out anyway, to get this merged as soon as possible. I'll send a separate PR with those.

NathanielHill commented:
@TooTallNate What's the process for a builder going from `canary` to a normal versioned release?

TooTallNate commented:
@NathanielHill Thank you for the PR, and sorry for taking so long to merge it. [CircleCI will be publishing a `canary` tag shortly](https://circleci.com/gh/zeit/now-builders/812).

We need to verify that all the `now-examples` demos work with the new tag and then we can promote it to a stable release.

honzajavorek commented:
There's a bug, missing `base64` import. I'll get it fixed in my PR.

NathanielHill commented:
@honzajavorek If you want to take the lead on implementing general WSGI support, I think that is what's most needed. There are several branches where people gave it a go, perhaps you can combine them and submit a nice PR. I believe that should cover it for most Python use cases: either a custom handler for really simple lambdas, or WSGI so you can drop in any framework :+1: 

This PR fixes everything I need for my *one* python lambda, so I'm unlikely to work on this anymore :man_shrugging: 

@TooTallNate No problem, glad to finally get it merged. Can't wait to update my `now.json` back to all official builders :ok_hand:

NathanielHill commented:
@honzajavorek Another thing you could have a look at is lambda size. My python lambda is a few hundred lines and ends up at ~8MB versus my similar JavaScript lambdas that are ~100kB even when including external libs.

honzajavorek commented:
A fix is here: https://github.com/zeit/now-builders/pull/226

@NathanielHill Thanks for the tips, I'd love to. Let's see how much time I have left - unfortunately it's just hobby work for me and there's only so many late night coding sessions I can spend on things 🙂

422158 commented:
> @honzajavorek Another thing you could have a look at is lambda size. My python lambda is a few hundred lines and ends up at ~8MB versus my similar JavaScript lambdas that are ~100kB even when including external libs.
@NathanielHill 
might be related to #201

NathanielHill commented:
@honzajavorek Forgot to mention #209 where it seems that things regressed from 3.6 -> 3.4 mysteriously. Not an issue for my project, but might be an easier one to tackle.
